File: clients/flutter/lib/src/watermark.dart
import 'error.dart';
import 'operations.dart';

/// The resolved overlay: a wire spec plus optional bytes to upload.
typedef ResolvedMark = ({Map<String, Object?> spec, List<int>? bytes});

/// An overlay passed to `TransformBuilder.composite` — a [Watermark] (image) or a
/// [Textmark] (text).
abstract interface class CompositeMark {
  /// Resolves the overlay into its wire spec, plus optional bytes to upload.
  ResolvedMark resolve();
}

/// An image overlay composited onto the result. Build it from a URL the server fetches (a
/// `String` or `Uri`) or from image bytes uploaded with the request (`List<int>`), then
/// chain the placement and appearance methods.
///
/// ```dart
/// st.transform(input)
///     .composite(Watermark('https://cdn.example.com/logo.png').resize(120).opacity(0.6))
///     .convert('webp');
/// ```
class Watermark implements CompositeMark {
  /// Creates an image watermark from [source]: a URL (`String`/`Uri`) the server fetches, or
  /// image bytes (`List<int>`) uploaded with the request.
  Watermark(Object source) {
    if (source is String) {
      _spec['url'] = source;
    } else if (source is Uri) {
      _spec['url'] = source.toString();
    } else if (source is List<int>) {
      _bytes = source;
    } else {
      throw const SharptownError(
          'Watermark(source): expected a URL String/Uri or image bytes (List<int>)');
    }
  }

  final Map<String, Object?> _spec = {'type': 'image'};
  List<int>? _bytes;

  /// Fits the overlay inside [width]×[height]. Either dimension is optional.
  Watermark resize([int? width, int? height]) {
    if (width != null) _spec['width'] = Operations.positiveInt(width, 'watermark width');
    if (height != null) _spec['height'] = Operations.positiveInt(height, 'watermark height');
    return this;
  }

  /// Sets the overlay width only.
  Watermark width(int value) {
    _spec['width'] = Operations.positiveInt(value, 'watermark width');
    return this;
  }

  /// Sets the overlay height only.
  Watermark height(int value) {
    _spec['height'] = Operations.positiveInt(value, 'watermark height');
    return this;
  }

  /// Rotates the overlay by [degrees].
  Watermark rotate(int degrees) {
    _spec['rotate'] = degrees;
    return this;
  }

  /// Sets the overlay opacity (0–1).
  Watermark opacity(num value) {
    _spec['opacity'] = Operations.range(value, 'watermark opacity', 0, 1);
    return this;
  }

  /// Sets the placement gravity (default `southeast`).
  Watermark gravity(String value) {
    _spec['gravity'] = value;
    return this;
  }

  /// Places the overlay at ([x], [y]) from the top-left instead of a gravity.
  Watermark offset(int x, int y) {
    _spec['x'] = x;
    _spec['y'] = y;
    return this;
  }

  /// Repeats the overlay across the whole image.
  Watermark tile([bool enabled = true]) {
    _spec['tile'] = enabled;
    return this;
  }

  /// Sets the Sharp blend mode (default `over`).
  Watermark blend(String mode) {
    _spec['blend'] = mode;
    return this;
  }

  @override
  ResolvedMark resolve() => (spec: _spec, bytes: _bytes);
}

/// A text overlay composited onto the result, rendered server-side. Pass it to
/// `TransformBuilder.composite`.
///
/// ```dart
/// st.transform(input)
///     .composite(Textmark('© Acme').size(48).color('white').rotate(-30).tile())
///     .convert('webp');
/// ```
class Textmark implements CompositeMark {
  /// Creates a text watermark.
  Textmark(String text) {
    _spec['text'] = text;
  }

  final Map<String, Object?> _spec = {'type': 'text'};

  /// Font size in pixels.
  Textmark size(int value) {
    _spec['size'] = Operations.positiveInt(value, 'textmark size');
    return this;
  }

  /// Text colour (any CSS colour).
  Textmark color(String value) {
    _spec['color'] = value;
    return this;
  }

  /// Font family.
  Textmark font(String value) {
    _spec['font'] = value;
    return this;
  }

  /// Font weight (e.g. `bold`).
  Textmark weight(String value) {
    _spec['weight'] = value;
    return this;
  }

  /// Background colour painted behind the text tile.
  Textmark background(String value) {
    _spec['background'] = value;
    return this;
  }

  /// Rotates the text by [degrees].
  Textmark rotate(int degrees) {
    _spec['rotate'] = degrees;
    return this;
  }

  /// Text opacity (0–1).
  Textmark opacity(num value) {
    _spec['opacity'] = Operations.range(value, 'textmark opacity', 0, 1);
    return this;
  }

  /// Placement gravity.
  Textmark gravity(String value) {
    _spec['gravity'] = value;
    return this;
  }

  /// Places the text at ([x], [y]) from the top-left.
  Textmark offset(int x, int y) {
    _spec['x'] = x;
    _spec['y'] = y;
    return this;
  }

  /// Repeats the text across the whole image.
  Textmark tile([bool enabled = true]) {
    _spec['tile'] = enabled;
    return this;
  }

  @override
  ResolvedMark resolve() => (spec: _spec, bytes: null);
}

File: packages/client/src/watermark.mjs
import { SharptownError } from './errors.mjs'
import { toInt, toPositiveInt, toRange } from './operations.mjs'

/**
 * @typedef {object} ResolvedMark
 * @property {object} spec The wire spec for the `composite` operation.
 * @property {Blob} [blob] Binary overlay to upload (for `File`/`Blob` sources).
 */

/**
 * An image overlay composited onto the transformed image. Pass it to
 * {@link TransformBuilder#composite}. The source is resolved by the server when it is a URL
 * (`string`, `URL`, or `Request`), or uploaded alongside the image when it is binary
 * (`File` or `Blob`). Configure placement and appearance with the chainable methods.
 *
 * @example
 * st.transform(file)
 *   .composite(new Watermark('https://cdn.example.com/logo.png').resize(120).opacity(0.6).gravity('southeast'))
 *   .convert('webp')
 */
export class Watermark {
  /** @type {File|Blob|Request|URL|string} */
  #source
  /** @type {Record<string, unknown>} */
  #spec = { type: 'image' }

  /** @param {File|Blob|Request|URL|string} source */
  constructor(source) {
    if (source == null || source === '') {
      throw new SharptownError('Watermark(source): a File, Blob, Request, URL or string is required')
    }
    this.#source = source
  }

  /**
   * Resizes the overlay, fitting inside `width`×`height`. Either dimension is optional.
   * @param {number} [width]
   * @param {number} [height]
   * @returns {this}
   */
  resize(width, height) {
    if (typeof width === 'object' && width !== null) {
      height = width.height
      width = width.width
    }
    if (width != null) this.#spec.width = toPositiveInt(width, 'watermark width')
    if (height != null) this.#spec.height = toPositiveInt(height, 'watermark height')
    return this
  }

  /** Sets the overlay width only. @param {number} value @returns {this} */
  width(value) {
    this.#spec.width = toPositiveInt(value, 'watermark width')
    return this
  }

  /** Sets the overlay height only. @param {number} value @returns {this} */
  height(value) {
    this.#spec.height = toPositiveInt(value, 'watermark height')
    return this
  }

  /** Rotates the overlay by degrees. @param {number} degrees @returns {this} */
  rotate(degrees) {
    this.#spec.rotate = toInt(degrees, 'watermark rotate')
    return this
  }

  /** Overlay opacity, `0`–`1`. @param {number} value @returns {this} */
  opacity(value) {
    this.#spec.opacity = toRange(value, 'watermark opacity', 0, 1)
    return this
  }

  /**
   * Placement gravity: `north`, `northeast`, `east`, `southeast`, `south`, `southwest`,
   * `west`, `northwest`, `center`. Defaults to `southeast`.
   * @param {string} value
   * @returns {this}
   */
  gravity(value) {
    this.#spec.gravity = String(value)
    return this
  }

  /** Absolute placement at `(x, y)` from the top-left, instead of a gravity. @param {number} x @param {number} y @returns {this} */
  offset(x, y) {
    this.#spec.x = toInt(x, 'watermark x')
    this.#spec.y = toInt(y, 'watermark y')
    return this
  }

  /** Repeats the overlay across the whole image. @param {boolean} [enabled=true] @returns {this} */
  tile(enabled = true) {
    this.#spec.tile = Boolean(enabled)
    return this
  }

  /** Sharp blend mode (`over` by default). @param {string} mode @returns {this} */
  blend(mode) {
    this.#spec.blend = String(mode)
    return this
  }

  /**
   * Resolves the overlay into a wire spec, plus a `Blob` to upload when the source is binary.
   * @returns {ResolvedMark}
   */
  resolve() {
    const spec = { ...this.#spec }
    const source = this.#source
    if (typeof source === 'string') {
      spec.url = source
    } else if (source instanceof URL) {
      spec.url = source.href
    } else if (typeof Request !== 'undefined' && source instanceof Request) {
      spec.url = source.url
    } else if (typeof Blob !== 'undefined' && source instanceof Blob) {
      return { spec, blob: source }
    } else {
      throw new SharptownError('Watermark(source): expected a File, Blob, Request, URL or string')
    }
    return { spec }
  }
}

/**
 * A text overlay composited onto the transformed image, rendered server-side. Pass it to
 * {@link TransformBuilder#composite}.
 *
 * @example
 * st.transform(file)
 *   .composite(new Textmark('© Acme').size(48).color('white').rotate(-30).tile())
 *   .convert('webp')
 */
export class Textmark {
  /** @type {Record<string, unknown>} */
  #spec = { type: 'text' }

  /** @param {string} text */
  constructor(text) {
    if (text == null) throw new SharptownError('Textmark(text): text is required')
    this.#spec.text = String(text)
  }

  /** Font size in pixels. @param {number} value @returns {this} */
  size(value) {
    this.#spec.size = toPositiveInt(value, 'textmark size')
    return this
  }

  /** Text colour (any CSS colour, e.g. `white`, `#fff`, `rgba(0,0,0,.5)`). @param {string} value @returns {this} */
  color(value) {
    this.#spec.color = String(value)
    return this
  }

  /** Font family. @param {string} value @returns {this} */
  font(value) {
    this.#spec.font = String(value)
    return this
  }

  /** Font weight (e.g. `bold`, `400`). @param {string|number} value @returns {this} */
  weight(value) {
    this.#spec.weight = String(value)
    return this
  }

  /** Background colour painted behind the text tile. @param {string} value @returns {this} */
  background(value) {
    this.#spec.background = String(value)
    return this
  }

  /** Rotates the text by degrees. @param {number} degrees @returns {this} */
  rotate(degrees) {
    this.#spec.rotate = toInt(degrees, 'textmark rotate')
    return this
  }

  /** Text opacity, `0`–`1`. @param {number} value @returns {this} */
  opacity(value) {
    this.#spec.opacity = toRange(value, 'textmark opacity', 0, 1)
    return this
  }

  /** Placement gravity (see {@link Watermark#gravity}). @param {string} value @returns {this} */
  gravity(value) {
    this.#spec.gravity = String(value)
    return this
  }

  /** Absolute placement at `(x, y)` from the top-left. @param {number} x @param {number} y @returns {this} */
  offset(x, y) {
    this.#spec.x = toInt(x, 'textmark x')
    this.#spec.y = toInt(y, 'textmark y')
    return this
  }

  /** Repeats the text across the whole image. @param {boolean} [enabled=true] @returns {this} */
  tile(enabled = true) {
    this.#spec.tile = Boolean(enabled)
    return this
  }

  /**
   * Resolves the text overlay into a wire spec.
   * @returns {ResolvedMark}
   */
  resolve() {
    return { spec: { ...this.#spec } }
  }
}
